v3-roadmap-and-design.md

```diff
@@ -0,0 +1,425 @@
+
+**`zarr.h5compat.Group`**
+
+Zarr-Python 2.* made an attempt to align its API with that of [h5py](https://docs.h5py.org/en/stable/index.html). With 3.0, we will relax this alignment in favor of providing an explicit compatibility module (`zarr.h5py_compat`). This module will expose the `Group` and `Dataset` APIs that map to Zarr-Python’s `Group` and `Array` objects. 
+
+### Creation API
+
+Zarr-Python 2.* bundles together the creation and serialization of Zarr objects. Zarr-Python 3.* will make it possible to create objects in memory separate from serializing them. This will specifically enable writing hierarchies of Zarr objects in a single batch step. For example:
+
+```python
+
+arr1 = Array(shape=(10, 10), path="foo/bar", dtype="i4", store=store)
+arr2 = Array(shape=(10, 10), path="foo/spam", dtype="f8", store=store)
+
+arr1.save()
+arr2.save()
+
+# or equivalently
+
+zarr.save_many([arr1 ,arr2])
+```
+
+### Plugin API
+
+Zarr V3 was designed to be extensible at multiple layers. Zarr-Python will support these extensions through a combination of [Abstract Base Classes](https://docs.python.org/3/library/abc.html) (ABCs) and [Entrypoints](https://packaging.python.org/en/latest/specifications/entry-points/). 
+
+**ABCs**
+
+Zarr V3 will expose Abstract base classes for the following objects:
+
+- `Store`, `ReadStore`, `ReadWriteStore`, `ReadListStore`, and `ReadWriteListStore`
+- `BaseArray`, `SynchronousArray`, and `AsynchronousArray`
+- `BaseGroup`, `SynchronousGroup`, and `AsynchronousGroup`
+- `Codec`, `ArrayArrayCodec`, `ArrayBytesCodec`, `BytesBytesCodec`
+
+**Entrypoints**
+
+Lots more thinking here but the idea here is to provide entrypoints for `data type`, `chunk grid`, `chunk key encoding`, `codecs`,  `storage_transformers` and `stores`.  These might look something like:
+
+```
+entry_points="""
+    [zarr.codecs]
+    blosc_codec=codec_plugin:make_blosc_codec
+    zlib_codec=codec_plugin:make_zlib_codec
+"""
+```
+
+### Python type hints and static analysis
+
+Target 100% Mypy coverage in 3.0 source. 
+
+### Observability
+
+A persistent problem in Zarr-Python is diagnosing problems that span many parts of the stack. To address this in 3.0, we will add a basic logging framework that can be used to debug behavior at various levels of the stack. We propose to add the separate loggers for the following namespaces:
+
+- `array`
+- `group`
+- `store`
+- `codec`
+
+These should be documented such that users know how to activate them and developers know how to use them when developing extensions.
+
+### Dependencies
+
+Today, Zarr-Python has the following required dependencies:
+
+```python
+dependencies = [
+    'asciitree',
+    'numpy>=1.20,!=1.21.0',
+    'fasteners',
+    'numcodecs>=0.10.0',
+]
+```
+
+What other dependencies should be considered?
+
+1. Attrs - Zarrita makes extensive use of the Attrs library
+2. Fsspec - Zarrita has a hard dependency on Fsspec. This could be easily relaxed though.
```

Comment:
(noting that fsspec has no dependencies of its own, but many of its implementations do)